core/src/package_store.rs

```diff
@@ -0,0 +1,559 @@
+                Ok(())
+            }
+        }
+    }
+
+    pub(crate) fn load_native_package(&mut self, pkg: Package) -> Result<(), CoreError> {
+        debug_assert_eq!(pkg.implementation, PackageImplementation::Native);
+        let name = &pkg.info.as_ref().name;
+        let entry = self.native_packages.entry(name.to_string());
+
+        match entry {
+            Entry::Occupied(_) => return Err(CoreError::OccupiedName(name.to_string())),
+            Entry::Vacant(entry) => entry.insert(pkg),
+        };
+        Ok(())
+    }
+
+    /// Gets information about a package with a given name
+    pub fn get_package_info(&self, name: &str) -> Option<Arc<PackageInfo>> {
+        self.native_packages
+            .get(name)
+            .or(self.standard_packages.get(&name.into()))
+            .map(|pkg| pkg.info.clone())
+    }
+
+    /// Borrow a vector with PackageInfo from every loaded package
+    pub fn get_all_package_info(&self) -> Vec<Arc<PackageInfo>> {
+        self.native_packages
+            .values()
+            .chain(self.standard_packages.values())
+            .chain(self.external_packages.values())
+            .map(|pkg| pkg.info.clone())
+            .collect::<Vec<_>>()
+    }
+
+    /// This gets the transform info for a given element and output format. If a native package
+    /// supplies a transform for that element, that will be returned and the output format returned
+    pub fn get_transform_info(
+        &self,
+        element_name: &str,
+        output_format: &OutputFormat,
+    ) -> Option<Transform> {
+        self.get_transform_to(element_name, output_format)
+            .map(|(transform, _)| transform)
+    }
+
+    /// Gets the `ArgInfo`s associated with an element targeting the given output format, if such
+    /// a transformation exists, otherwise generates an `MissingTransform` error. This is intended
+    /// for use in `collect_(parent/module)_arguments` to reduce repeated code.
+    pub(crate) fn get_args_info(
+        &self,
+        element_name: &str,
+        output_format: &OutputFormat,
+    ) -> Result<Vec<ArgInfo>, CoreError> {
+        self.get_transform_info(element_name, output_format)
+            .map(|info| info.arguments)
+            .ok_or(CoreError::MissingTransform(
+                element_name.to_string(),
+                output_format.0.to_string(),
+            ))
+    }
+
+    /// Gets the transform and package the transform is in, for a transform from a specific element
+    /// to a specific output format. Returns None if no such transform exists
+    pub fn get_transform_to(
+        &self,
+        element_name: &str,
+        output_format: &OutputFormat,
+    ) -> Option<(Transform, Package)> {
+        self.transforms
+            .get(element_name)
+            .and_then(|t| t.find_transform_to(output_format))
+            .cloned()
+    }
+
+    pub(crate) fn get_missing_packages(
+        &mut self,
+        arc_mutex: Arc<Mutex<Self>>,
```

Comment:
We need the arc_mutex since we need to clone the reference to store in the tasks. We can use that to get a lock, but since we do already have a lock when calling this, it seems bad to just drop it and then re-aquire it. Even though it isn't elegant, I think this has to do. Changed the name to `generate_resolve_tasks`.